### PR TITLE
Handle file upload future completed exceptionally for Java-based TM

### PR DESCRIPTION
**Pull request by davidh44**
## Motivation and Context

During a file upload with the Java-based TransferManager, if an outage occurs and the upload is paused, the resume token is not saved

## Modifications

Add a check to see if the upload future was completed exceptionally

## Testing

Tested end-to-end locally by shutting off wifi during upload
[Tests](https://github.com/aws/aws-sdk-java-v2/commit/7ce6c2709f795400727fab3df71efbd2fbade075)

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/aws-sdk-java-v2/blob/master/CONTRIBUTING.md) document
- [ ] Local run of `mvn install` succeeds
- [ ] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [ ] I have added tests to cover my changes
- [ ] All new and existing tests passed
- [ ] I have added a changelog entry. Adding a new entry must be accomplished by running the `scripts/new-change` script and following the instructions. Commit the new file created by the script in `.changes/next-release` with your changes.
- [ ] My change is to implement 1.11 parity feature and I have updated [LaunchChangelog](https://github.com/aws/aws-sdk-java-v2/blob/master/docs/LaunchChangelog.md)

## License

- [ ] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/model/CrtFileUpload.java [hunk 1]
```diff
@@ -57,7 +61,14 @@ public ResumableFileUpload pause() {
 
     private ResumableFileUpload doPause() {
         File sourceFile = request.source().toFile();
-        if (completionFuture.isDone()) {
+
+        boolean futureCompletedExceptionally = completionFuture.isCompletedExceptionally();
```
Comment (joviegas) [inline services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/model/CrtFileUpload.java:65]: Can we write mock test to test the same ?

Comment (davidh44) [reply services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/model/CrtFileUpload.java:65]: added test

File: services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/model/CrtFileUpload.java [hunk 2]
```diff
@@ -80,15 +88,24 @@ private ResumableFileUpload doPause() {
         }
 
         completionFuture.cancel(true);
-        // Upload hasn't started yet, or it's a single object upload
         if (token == null) {
+            // TODO - remove once CRT handles future completed exceptionally to return ResumeToken
```
Comment (joviegas) [inline services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/model/CrtFileUpload.java:92]: For this TODO do we have a follow up issue or task on CRT ?

File: services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/model/CrtFileUpload.java [hunk 3]
```diff
@@ -22,17 +22,21 @@
...
 import software.amazon.awssdk.transfer.s3.model.UploadFileRequest;
 import software.amazon.awssdk.transfer.s3.progress.TransferProgress;
 import software.amazon.awssdk.utils.Lazy;
+import software.amazon.awssdk.utils.Logger;
 import software.amazon.awssdk.utils.ToString;
 import software.amazon.awssdk.utils.Validate;
 
 @SdkInternalApi
 public final class CrtFileUpload implements FileUpload {
+    private static final Logger log = Logger.loggerFor(S3TransferManager.class);
```
Comment (joviegas) [inline services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/model/CrtFileUpload.java:38]: Logger for class should be CrtFileUpload

Comment (davidh44) [reply services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/model/CrtFileUpload.java:38]: Updated, we have a handful of classes that use S3TransferManager for logger, not sure if that's intentional

File: services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/model/CrtFileUpload.java (hunk 2 above)
Comment (davidh44) [reply services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/model/CrtFileUpload.java:92]: Yes ticket was cut to CRT

File: services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/model/CrtFileUpload.java [hunk 4]
```diff
@@ -80,15 +88,24 @@ private ResumableFileUpload doPause() {
...
+            }
+
             return ResumableFileUpload.builder()
                                       .fileLastModified(fileLastModified)
                                       .fileLength(sourceFile.length())
                                       .uploadFileRequest(request)
                                       .build();
         }
 
+        log.debug(() -> "The upload was successfully paused and a ResumeToken was returned.");
```
[the remark this answers is not in the document]
Comment (davidh44) [reply services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/model/CrtFileUpload.java:108]: Removed for now, will be helpful to include after CRT fix, in order to differentiate between future completed exceptionally, as in DefaultFileUpload

**Review by joviegas — APPROVED** [review]

**Pull request merged**